src/main/java/akka/stream/contrib/DirectoryChanges.java

```diff
@@ -0,0 +1,213 @@
+   */
+  public DirectoryChanges(Path directoryPath, FiniteDuration pollInterval, int maxBufferSize) {
+    this.directoryPath = directoryPath;
+    this.pollInterval = pollInterval;
+    this.maxBufferSize = maxBufferSize;
+  }
+
+  @Override
+  public SourceShape<Pair<Path, Change>> shape() {
+    return shape;
+  }
+
+  @Override
+  public Attributes initialAttributes() {
+    return defaultAttributes;
+  }
+
+  @Override
+  public GraphStageLogic createLogic(Attributes inheritedAttributes) {
+    if(!Files.exists(directoryPath)) throw new IllegalArgumentException("The path: '" + directoryPath + "' does not exist");
+    if(!Files.isDirectory(directoryPath)) throw new IllegalArgumentException("The path '" + directoryPath + "' is not a directory");
+
+    try {
+      return new TimerGraphStageLogic(shape) {
+        private final Queue<Pair<Path, Change>> buffer = new ArrayDeque<>();
+        private final WatchService service = directoryPath.getFileSystem().newWatchService();
+        private final WatchKey watchKey = directoryPath.register(
+          service,
+          new WatchEvent.Kind<?>[] { ENTRY_CREATE, ENTRY_MODIFY, ENTRY_DELETE, OVERFLOW },
+          // this is com.sun internal, but the service is useless on OSX without it
+          SensitivityWatchEventModifier.HIGH
+        );
+
+        {
+          setHandler(out, new AbstractOutHandler(){
+
+            @Override
+            public void onPull() throws Exception {
+              if (!buffer.isEmpty()) {
+                pushHead();
+              } else {
+                doPoll();
+                if (!buffer.isEmpty()) {
+                  pushHead();
+                } else {
+                  schedulePoll();
+                }
+              }
+            }
+          });
+        }
+
+        @Override
+        public void onTimer(Object timerKey) {
+          if (!isClosed(out)) {
+            doPoll();
+            if (!buffer.isEmpty()) {
+              pushHead();
+            } else {
+              schedulePoll();
+            }
+          }
+        }
+
+        @Override
+        public void postStop() {
+          try {
+            if (watchKey.isValid()) watchKey.cancel();
+            service.close();
+          } catch (Exception ex) {
+            // Remove when #21168 is in a release
+            throw new RuntimeException(ex);
+          }
+        }
+
+        private void pushHead() {
+          final Pair<Path, Change> head = buffer.poll();
+          if (head != null) {
+            push(out, head);
+          }
+        }
+
+        private void schedulePoll() {
+          scheduleOnce("poll", pollInterval);
+        }
+
+        private void doPoll() {
+          try {
+            for (WatchEvent<?> event: watchKey.pollEvents()) {
+              final WatchEvent.Kind<?> kind = event.kind();
+
+              if (OVERFLOW.equals(kind)) {
+                failStage(new RuntimeException("Overflow from watch service: '" + directoryPath + "'"));
```

Comment:
Add a comment explaining that technically one may choose to ignore the overflow instead - depends what you need.
No need to change impl, but add comment to explain for readers we're aware